pennylane/transforms/mitigate.py

```diff
@@ -12,15 +12,346 @@
+
+    Args:
+        circuit (QuantumTape): the circuit to be folded
+        scale_factor (float): Scale factor :math:`\lambda` determining :math:`n` and :math:`s`
+
+    Returns:
+        QuantumTape: Folded circuit
+
+    **Example:**
+
+    .. code-block:: python
+
+        x = np.arange(6)
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(x[0], wires=0)
+            qml.RY(x[1], wires=1)
+            qml.RZ(x[2], wires=2)
+            qml.CNOT(wires=(0,1))
+            qml.CNOT(wires=(1,2))
+            qml.RX(x[3], wires=0)
+            qml.RY(x[4], wires=1)
+            qml.RZ(x[5], wires=2)
+            qml.expval(qml.PauliZ(0) @ qml.PauliZ(1) @ qml.PauliZ(2))
+
+    Setting ``scale_factor = 1`` does not affect the circuit:
+
+    >>> folded_tape = qml.transforms.fold_global(tape, 1)
+    >>> print(qml.drawer.tape_text(folded_tape, decimals=1))
+    0: ──RX(0.0)─╭●──RX(3.0)──────────┤ ╭<Z@Z@Z>
+    1: ──RY(1.0)─╰X─╭●────────RY(4.0)─┤ ├<Z@Z@Z>
+    2: ──RZ(2.0)────╰X────────RZ(5.0)─┤ ╰<Z@Z@Z>
+
+    Setting ``scale_factor = 2`` results in the partially folded circuit :math:`U (L^\dagger_d L^\dagger_{d-1} .. L^\dagger_s) (L_s .. L_d)`
+    with :math:`s = \lfloor \left(1 \mod 2 \right) d/2 \rfloor = 4` since the circuit is composed of :math:`d=8` gates.
+
+    >>> folded_tape = qml.transforms.fold_global(tape, 2)
+    >>> print(qml.drawer.tape_text(folded_tape, decimals=1))
+    0: ──RX(0.0)─╭●──RX(3.0)──RX(3.0)†──RX(3.0)──────────────────┤ ╭<Z@Z@Z>
+    1: ──RY(1.0)─╰X─╭●────────RY(4.0)───RY(4.0)†─╭●──╭●──RY(4.0)─┤ ├<Z@Z@Z>
+    2: ──RZ(2.0)────╰X────────RZ(5.0)───RZ(5.0)†─╰X†─╰X──RZ(5.0)─┤ ╰<Z@Z@Z>
+
+    Setting ``scale_factor = 3`` results in the folded circuit :math:`U (U^\dagger U)`.
+
+    >>> folded_tape = qml.transforms.fold_global(tape, 3)
+    >>> print(qml.drawer.tape_text(folded_tape, decimals=1))
+    0: ──RX(0.0)─╭●──RX(3.0)──RX(3.0)†───────────────╭●─────────RX(0.0)†──RX(0.0)─╭●──RX(3.0)──────────┤╭<Z@Z@Z>
+    1: ──RY(1.0)─╰X─╭●────────RY(4.0)───RY(4.0)†─╭●──╰X†────────RY(1.0)†──RY(1.0)─╰X─╭●────────RY(4.0)─┤├<Z@Z@Z>
+    2: ──RZ(2.0)────╰X────────RZ(5.0)───RZ(5.0)†─╰X†──RZ(2.0)†──RZ(2.0)──────────────╰X────────RZ(5.0)─┤╰<Z@Z@Z>
+    """
+
+    if scale_factor < 1.0:
+        raise AttributeError("scale_factor must be >= 1")
+    assert scale_factor >= 1.0
+
+    # TODO: simplify this - currently just a workaround, problem is when tape contains adjoint(op)
+    def qfunc(op):
+        copy(op).queue()
+
+    # Generate base_circuit without measurements
+    # Treat all circuits as lists of operations, build new tape in the end
+
+    base_ops = circuit.expand().copy(copy_operations=True).operations
+
+    def _divmod(a, b):
+        """Performs divmod but in an all-interface compatible manner"""
+        out1 = qml.math.floor(a / b)
+        out2 = a - out1 * b
+        return int(out1), int(out2)
```

Comment:
Yes indeed. I was thinking about this but was not sure how to handle the `int()` returns since this would break differentiablility?

Reply:
Could I just add this to batch_dispatch?

Reply:
Sorry, I think I wasn't clear; I meant, rather than having this inside the transform, could it be moved outside the function? It looks like the closure isn't needed here?

Reply:
Ah right, gotcha! I didnt even realize this was still inside the transform